cmd/juno/juno.go

```diff
@@ -0,0 +1,123 @@
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/NethermindEth/juno/internal/juno"
+	"github.com/NethermindEth/juno/internal/utils"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+const greeting = `
+       _                    
+      | |                   
+      | |_   _ _ __   ___   
+  _   | | | | | '_ \ / _ \  
+ | |__| | |_| | | | | (_) |  
+  \____/ \__,_|_| |_|\___/  
+
+Juno is a Go implementation of a StarkNet full node client made with ❤️ by Nethermind.
+
+`
+
+const (
+	configF    = "config"
+	verbosityF = "verbosity"
+	rpcPortF   = "rpc-port"
+	metricsF   = "metrics"
+	dbPathF    = "db-path"
+	networkF   = "network"
+	ethNodeF   = "eth-node"
+
+	defaultConfig    = ""
+	defaultVerbosity = "info"
+	defaultRpcPort   = uint16(6060)
+	defaultMetrics   = false
+	defaultDbPath    = ""
+	defaultNetwork   = utils.GOERLI
+	defaultEthNode   = ""
+
+	configFlagUsage    = "The yaml configuration file."
+	verbosityFlagUsage = "Verbosity of the logs. Options: debug, info, warn, error, dpanic, " +
+		"panic, fatal."
+	rpcPortUsage = "The port on which the RPC server will listen for requests. " +
+		"Warning: this exposes the node to external requests and potentially DoS attacks."
+	metricsUsage = "Enables the metrics server and listens on port 9090."
+	dbPathUsage  = "Location of the database files."
+	networkUsage = "Available StarkNet networks. Options: 0 = goerli and 1 = mainnet"
+	ethNodeUsage = "The Ethereum endpoint to synchronise with. " +
+		"If unset feeder gateway will be used."
+)
+
+var (
+	StarkNetNode juno.StarkNetNode
+	cfgFile      string
+)
+
+func NewCmd(newNodeFn juno.NewStarkNetNodeFn, quit <-chan os.Signal) *cobra.Command {
```

Comment:
Rather than passing in a setup function as a parameter, would it be slightly more idiomatic to define a `LoadConfig(cfg *juno.Config) error` function in the `StarknetNode` interface? Then we can pass in a `node StarknetNode` struct and call `node.LoadConfig(cfg)` on line 98.

I don't see any tangible benefits to doing it this way, so I'll definitely approve (the PR is awesome in many respects), but I am curious to hear your thoughts.

Reply:
Interfaces need to represent a conceptual idea, that is why if you look at the standard lib's `Reader` and `Writer` interface they are just one function. So my thinking behind the `StarkNetNode` interface was: "What is the minimum I need the StarkNetNode to do?" The immediate answer was that I should be able to run and stop a node.

The reason for passing a new node function was mainly a dependency injection for the tests. I personally wouldn't like to include `LoadConfig(cfg *juno.Config) error` to the interface because it implies that calling the new node function is not sufficient to run/start the node (this defeats the intention behind the `New()`). We don't want to have partially initialised structs that are being passed around. This can lead to panic when things are not fully initialised.

Also, initially, I did think about passing the `StarkNetNode` as a parameter but it didn't work because that meant I needed the config before calling `NewCmd()` which defeated the purpose of `NewCmd()`. So your suggestion does offer a solution. However, we need to think about other scenarios of how the `StarkNetNode` can be used and I feel `LoadConfig(cfg *juno.Congig) error` in the interface is superfluous and a more idiomatic way is to have `New()` doing all the initialisation.